File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_Stoch_D_K_ThreeAscendingFromOversold.java
package com.kishlaly.ta.analyze.tasks.blocks.two;

import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.model.indicators.IndicatorJava;
import com.kishlaly.ta.model.indicators.StochJava;
import com.kishlaly.ta.utils.LogJava;

import java.util.List;

import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.model.BlockResultCodeJava.STOCH_D_K_NOT_ASCENDING_FROM_OVERSOLD_SCREEN_2;
import static com.kishlaly.ta.analyze.tasks.ThreeDisplaysJava.Config.STOCH_OVERSOLD;

public class Long_ScreenTwo_Stoch_D_K_ThreeAscendingFromOversold implements ScreenTwoBlockJava {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        List<StochJava> screen_2_Stochastic = (List<StochJava>) screen.indicators.get(IndicatorJava.STOCH);
        StochJava stoch3 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 3);
        StochJava stoch2 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 2);
        StochJava stoch1 = screen_2_Stochastic.get(screen_2_Stochastic.size() - 1);

        // the third or second from the end of %K below STOCH_OVERSOLD, and the very last above the first
        boolean isOversoldK = (stoch3.getSlowK() <= STOCH_OVERSOLD || stoch2.getSlowK() <= STOCH_OVERSOLD)
                && (stoch1.getSlowK() > stoch3.getSlowK());

        // the third or second from the end of %D below STOCH_OVERSOLD, and the very last above both
        boolean isOversoldD = (stoch3.getSlowD() <= STOCH_OVERSOLD || stoch2.getSlowD() <= STOCH_OVERSOLD)
                && (stoch1.getSlowD() > stoch2.getSlowD())
                && (stoch1.getSlowD() > stoch3.getSlowD());

        if (!isOversoldK || !isOversoldD) {
            LogJava.recordCode(STOCH_D_K_NOT_ASCENDING_FROM_OVERSOLD_SCREEN_2, screen);
            LogJava.addDebugLine("Stochastic does not rise from oversold " + STOCH_OVERSOLD + ". %D: " + isOversoldD + "; %K: " + isOversoldK);
            return new BlockResultJava(screen.getLastQuote(), STOCH_D_K_NOT_ASCENDING_FROM_OVERSOLD_SCREEN_2);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/two/Long_ScreenTwo_Stoch_D_K_ThreeAscendingFromOversold.kt
package com.kishlaly.ta.analyze.tasks.blocks.two

import com.kishlaly.ta.analyze.tasks.ThreeDisplays
import com.kishlaly.ta.analyze.tasks.ThreeDisplays.Config.STOCH_OVERSOLD
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.model.indicators.Stochastic
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.SymbolDataUtils

class Long_ScreenTwo_Stoch_D_K_ThreeAscendingFromOversold : ScreenTwoBlock {
    override fun check(screen: SymbolData): BlockResult {
        val stoch = SymbolDataUtils(screen, Stochastic::class.java)

        // the third or second from the end of %K below STOCH_OVERSOLD, and the very last above the third
        val isOversoldK = (stoch.last(3).slowK <= STOCH_OVERSOLD || stoch.last(2).slowK <= STOCH_OVERSOLD)
                && stoch.last().slowK > stoch.last(3).slowK

        // the third or second from the end of %D below STOCH_OVERSOLD, and the very last above both
        val isOversoldD = (stoch.last(3).slowD <= STOCH_OVERSOLD || stoch.last(2).slowD <= STOCH_OVERSOLD)
                && stoch.last().slowD > stoch.last(2).slowD && stoch.last().slowD > stoch.last(3).slowD

        if (!isOversoldK || !isOversoldD) {
            Log.recordCode(BlockResultCode.STOCH_D_K_NOT_ASCENDING_FROM_OVERSOLD_SCREEN_2, screen)
            Log.addDebugLine("Stochastic does not rise from oversold " + ThreeDisplays.Config.STOCH_OVERSOLD + ". %D: " + isOversoldD + "; %K: " + isOversoldK)
            return BlockResult(
                screen.lastQuote, BlockResultCode.STOCH_D_K_NOT_ASCENDING_FROM_OVERSOLD_SCREEN_2
            )
        }

        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}